Gha check branch functions

hansvancalster opened the pull request:
- [x] extra checks for protocol specific `NEWS.md` added to check_structure()
- [x] updated `NEWS.md` template files
- [x] new function `get_version_number()`
- [x] new function `update_news_release()`
- [x] new function `update_version_number()`
- [x] new function `set_tags()`

hansvancalster commented:
To do:

- [ ] in check_frontmatter() check if version number needs to be updated with update_version_number()
- [ ] adapt workflow in protocolsource repo

ElsLommelen commented on R/check_structure.R:
```diff
@@ -183,3 +186,76 @@ check_file <- function(filename, x, files_template, path_to_template) {
+    version_number <- yml$version_number
+
+    if (!any(str_detect(version_number, headings_text))) {
+      return(
+        x$add_error(
+          "An entry is missing in NEWS.md for this version of the protocol")
+      )
```
```suggestion
        x$add_error(
          "An entry is missing in NEWS.md for this version of the protocol")
```

ElsLommelen reviewed (commented):
For this review I tested all functions that are needed for the automation, and while doing this, I also notices some issues that actually refer to earlier code:

- `update_zenodo()`
    - [x] despite this is an internal function, the documentation for this function is published externally (i.e. accessible with `?update_zenodo`)  (I accidentally noticed it for this function and I didn't check it for other internal functions.)
    - [ ] by simply running this function I somehow managed to have twice the same name in the `.zenodo.json` file, once as a creator without the affiliation added (and orcid with prefix `https://orcid.org/`), and once as author of a protocol with the affiliation added and just the orcid without website.  As this file will only be added in a final step in the PR, I suppose it will not be checked anymore and this error will thus escape the attention.  So maybe author names should be checked more carefully in this function?  (To reproduce the described case: just run `create_protocol()` in project protocolhelper and make sure to add one of the package authors in the protocol.)

(Other functions look fine or I already added comments, as far as I can judge by just running them locally in RStudio.)

hansvancalster commented:
This reminds me of an idea @ThierryO had with regard to `INBOmd` package. I think the idea was to have some standard file with author names / orcids to avoid having to type author names (this was in the context of INBOmd templates with interactive questions to fill in the YAML metadata). 

Can we for instance generate a file containing all INBO personnel that has an orcid? That would make it relatively easy to write checks and catch typos.

ThierryO commented:
You could use [`checklist::orcid2person()`](https://inbo.github.io/checklist/reference/orcid2person.html) get the name and email from ORCID.

ElsLommelen commented:
Apparently there is a package rorcid that [allows to retrieve information](https://www.pauloldham.net/introduction-to-orcid-with-rorcid/#:~:text=We%20have%20an%20ORCID%20identifier%20and%20we%20want%20to%20retrieve%20information%20such%20as%20publications%20or%20funding%20information.).  I'm not sure it also allows to get the author name.

ThierryO commented:
[`checklist::orcid2person()`](https://github.com/inbo/checklist/blob/84cd5114dbc8125c6c99721716fca16e5a92375a/R/utils.R#L95) uses `rorcid` ;-)

hansvancalster commented:
> You could use [`checklist::orcid2person()`](https://inbo.github.io/checklist/reference/orcid2person.html) get the name and email from ORCID.

@ElsLommelen the protocolhelper package already has an ORCID_TOKEN secret which you can use for authentication with this function inside a GHA.

hansvancalster commented:
> I copied `.zenodo.json` from protocolsource, so it should be conform, and I used a protocol in which I had added author `Hans Van Calster` with the correct orcid (which I made some months ago with `create_protocol()`). 

Could it be you accidentally copied `.zenodo.json` from protocolhelper instead of protocolsource? That would explain where the `https://orcid.org/` prefix came from.

>As far as I remember, `check_frontmatter()` didn't give any errors on this, but right now this function doesn't run due to authentication problems with ssh (no idea why it is trying to authenticate anyway while running locally, but it might have to do with the fact that I had to renew my PAT today, while I use ssh to authenticate to github). So this protocol seems to be conform, but still `update_zenodo()` adds your name twice.

Can you check if it now works as expected with current state of this branch?

> This same error would occur with a simply typo in one of the names, so maybe it should be better to just use the orcid to test for author similarity?

This is now implemented in `update_zenodo()`. I also added a function `validate_orcid()`, which is used by `check_frontmatter` to make sure the ORCIDs are valid.

ElsLommelen commented:
> Could it be you accidentally copied `.zenodo.json` from protocolhelper instead of protocolsource? That would explain where the `https://orcid.org/` prefix came from.

In my first trial (post from 14 july) I indeed did, but in my second trial (15 july) which I refer to here, I copied `.zenodo.json` from protocolsource.  In this case the `https://orcid.org/` was of course absent, so the orcids were identical, but I still got the problem that an author is added twice if the name is typed differently (in this case `Van Calster, Hans` opposite to `Hans Van Calster`, but just a typo would likely give the same result).  I tried it again right now, after updating, but I still get the same result: when running `update_zenodo()` the name is added twice without dropping an error or giving a message.

> Can you check if it now works as expected with current state of this branch?

No, apparently `check_frontmatter()` cannot be run locally anymore (due to the switch to the main branch for one of the tests?), and I am not yet completely ready to test it in docker: there seems to be a problem with a date format in the protocol that I am testing there (sfp-401-nl), dropped by function `format_ISO8601` because the character string is not in a standard unambiguous format.  So this has to be solved before I can continue testing, I'm afraid (either in the `check_frontmatter()` function or in the protocol, but you may want these cases to be captured in `check_frontmatter()` as they force a user to solve this one error with ambiguous message before they can continue).

Also, as `check_frontmatter()` cannot be run locally, you may want to consider making it an internal function.

ElsLommelen commented:
Ok, I already figured out the source of the date problem: the protocol has date ```"`r Sys.Date()`"```, which first is used in the function `as.Date()` in `check_frontmatter()`, and this gives:
```
> as.Date("`r Sys.Date()`")
Error in charToDate(x) : 
  character string is not in a standard unambiguous format
```
I'll change the date in my protocolsource clone to be able to continue testing, but you may consider to catch this exception in `check_frontmatter()`, as this invalid input is likely to be given once in a while.

hansvancalster commented:
> ... (in this case `Van Calster, Hans` opposite to `Hans Van Calster`, but just a typo would likely give the same result). I tried it again right now, after updating, but I still get the same result: when running `update_zenodo()` the name is added twice without dropping an error or giving a message.

I tried to reproduce this in [2132157](https://github.com/inbo/protocolhelper/pull/73/commits/213215792b46216efaf0e9e440f3e70211ace59c), but it seems to be working as expected.

> No, apparently `check_frontmatter()` cannot be run locally anymore (due to the switch to the main branch for one of the tests?),
> Also, as `check_frontmatter()` cannot be run locally, you may want to consider making it an internal function.

This is due to the use of `checklist::clean_git()` inside `protocolhelper:::get_version_number()`, which is called in one of the checks in `check_frontmatter()`. It's not really that it cannot be run locally though. It's just that you need to make sure that the remote is up to date and the protocolsource' branching model is correctly followed (you can look at the code in `test_check_frontmatter.R` for an example of unit test where this branching model is used and remote/local are kept synchronized). You must make sure that in your docker test environment, there is a remote counterpart (on your fork). The function `checklist::new_branch()` should be used to start a new branch (alternatively, consult the code from that function to do the equivalent with git commands).  

> and I am not yet completely ready to test it in docker: there seems to be a problem with a date format in the protocol that I am testing there (sfp-401-nl), dropped by function `format_ISO8601` because the character string is not in a standard unambiguous format. So this has to be solved before I can continue testing, I'm afraid (either in the `check_frontmatter()` function or in the protocol, but you may want these cases to be captured in `check_frontmatter()` as they force a user to solve this one error with ambiguous message before they can continue).

Solved in [77e62e8](https://github.com/inbo/protocolhelper/pull/73/commits/77e62e81343a052369e9a172ff43a271be05125a)

ElsLommelen commented:
> This is due to the use of `checklist::clean_git()` inside `protocolhelper:::get_version_number()`, which is called in one of the checks in `check_frontmatter()`. It's not really that it cannot be run locally though. It's just that you need to make sure that the remote is up to date and the protocolsource' branching model is correctly followed (you can look at the code in `test_check_frontmatter.R` for an example of unit test where this branching model is used and remote/local are kept synchronized). You must make sure that in your docker test environment, there is a remote counterpart (on your fork). The function `checklist::new_branch()` should be used to start a new branch (alternatively, consult the code from that function to do the equivalent with git commands).

Finally I can run it without problems in RStudio, but not in docker (where I simulate the GitHub environment).  Are you sure `checklist::clean_git()` can be used in GHA?  I had a quick look in checklist, and it seems this function is nowhere used in functions that are used in Docker/GHA.  Also: I see no reason to use it in GHA, as the 'workspace' is always clean over there (unless the GHA itself creates a mess ;-) ).  For some reason this function fails there, but this might just have to do with the fact that this environment differs from a local environment?  Is it possible to skip `checklist::clean_git()` (if !interactive())?

hansvancalster commented:
> For some reason this function fails there, but this might just have to do with the fact that this environment differs from a local environment? Is it possible to skip `checklist::clean_git()` (if !interactive())?

I can certainly do that, but I wonder if the problem on GHA has to do with [authentication](https://docs.github.com/en/actions/security-guides/automatic-token-authentication). Does the workflow have a GITHUB_TOKEN?

ElsLommelen commented:
> > For some reason this function fails there, but this might just have to do with the fact that this environment differs from a local environment? Is it possible to skip `checklist::clean_git()` (if !interactive())?
> 
> I can certainly do that, but I wonder if the problem on GHA has to do with [authentication](https://docs.github.com/en/actions/security-guides/automatic-token-authentication). Does the workflow have a GITHUB_TOKEN?

Yes it has (named GITHUB_PAT, but this name has to be passed in the yaml file anyway, so the exact name shouldn't make a difference).  Do I have to pass this token to the function `check_frontmatter()`?  I don't see an argument for it.
But honestly I don't think this can be the problem: the first step in my procedure is cloning the branch to the docker environment, and there the authentication works fine... (Without authenticating here, I simply wouldn't be able to get to the content of the branch to test it.)

hansvancalster commented:
> Yes it has (named GITHUB_PAT, but this name has to be passed in the yaml file anyway, so the exact name shouldn't make a difference). Do I have to pass this token to the function `check_frontmatter()`? I don't see an argument for it. But honestly I don't think this can be the problem: the first step in my procedure is cloning the branch to the docker environment, and there the authentication works fine... (Without authenticating here, I simply wouldn't be able to get to the content of the branch to test it.) Also, `checklist::git_clean()` seems not to use any token to authenticate, and this function is the culprit that causes the error.

The `checklist::clean_git()` function depends on several functions from the `gert` package which require authentication.  Anywat, the `clean_git` stuff is now skipped if not interactive [42b706e](https://github.com/inbo/protocolhelper/pull/73/commits/42b706e50741c12eb9587356329b78da2f378f99). So this problem is now bypassed.
